polymorphic.ts

```diff
@@ -0,0 +1,142 @@
+import { JsonValue, Serializable } from "./serializable.ts";
+
+const POLYMORPHIC_RESOLVER_MAP = new Map<
+  unknown,
+  (input: string | JsonValue | Object) => Serializable
+>();
+
+export function registerPolymorphicResolver(
+  classPrototype: unknown,
+  resolver: (input: string | JsonValue | Object) => Serializable,
+): void {
+  POLYMORPHIC_RESOLVER_MAP.set(classPrototype, resolver);
+}
+
+// Parent -> key -> value -> initializer
+const POLYMORPHIC_SWITCH_MAP = new Map<
+  unknown,
+  Map<string | symbol, Map<unknown, () => Serializable>>
+>();
+
+export function registerPolymorphicSwitch(
+  parentPrototype: unknown,
+  propertyName: string | symbol,
+  propertyValue: unknown,
+  initializerFunction: () => Serializable,
+): void {
+  // Assert parentPrototype != null
+
+  let classMap = POLYMORPHIC_SWITCH_MAP.get(parentPrototype);
+  if (!classMap) {
+    POLYMORPHIC_SWITCH_MAP.set(parentPrototype, new Map());
+    classMap = POLYMORPHIC_SWITCH_MAP.get(parentPrototype);
+  }
+
+  let propertyKeyMap = classMap?.get(propertyName);
+  if (!propertyKeyMap) {
+    classMap?.set(propertyName, new Map());
+    propertyKeyMap = classMap?.get(propertyName);
+  }
+
+  propertyKeyMap?.set(propertyValue, initializerFunction);
+}
+
+function resolveSwitchMap(
+  classPrototype: unknown,
+  input: string | JsonValue | Object,
+): Serializable | null {
+  const classMap = POLYMORPHIC_SWITCH_MAP.get(classPrototype);
+
+  if (!classMap) {
+    throw new Error(`Could not determine classMap for ${classPrototype}`);
+  }
+
+  const inputObject = typeof input === "string" ? JSON.parse(input) : input;
+
+  for (const [propertyKey, valueMap] of classMap.entries()) {
+    const value = inputObject[propertyKey];
+    const resolver = valueMap.get(value);
+    if (resolver) {
+      return resolver();
+    }
+  }
+
+  return null;
+}
+
+export function serializePolymorphicClass<T extends Serializable>(
+  classPrototype: { prototype: T },
+  input: string | JsonValue | Object,
+): T {
+  return (resolvePolymorphicClass(classPrototype, input) as T).fromJson(input);
+}
+
+function resolvePolymorphicClass(
+  classPrototype: unknown,
+  input: string | JsonValue | Object,
+): unknown {
+  const classResolver = POLYMORPHIC_RESOLVER_MAP.get(classPrototype);
+  if (classResolver) {
+    return classResolver(input);
+  }
+
+  const resolvedClass = resolveSwitchMap(classPrototype, input);
+
+  if (resolvedClass) {
+    return resolvedClass;
+  }
+
+  throw new Error(`Could not resolve class for ${classPrototype}`);
+}
+
+// @PolymorphicResolver method decorator
```

Comment:
because these are top-level we should follow the deno style guide and use `/** comment */`
```suggestion
/** @PolymorphicResolver method decorator */
```